src/install/install.ts

```diff
@@ -0,0 +1,137 @@
+ * The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+
+// <reference path="../../custom_typings/bower.d.ts" />
+// <reference path="../../custom_typings/bower-json.d.ts" />
+// <reference path="../../custom_typings/bower-logger.d.ts" />
+
+import * as bower from 'bower';
+import * as path from 'path';
+import * as logging from 'plylog';
+import {read as readBowerJson} from 'bower-json';
+
+import defaultBowerConfig = require('bower/lib/config');
+import BowerLogger = require('bower-logger');
+import StandardRenderer = require('bower/lib/renderers/StandardRenderer');
+import BowerProject = require('bower/lib/core/Project');
+
+const logger = logging.getLogger('cli.install');
+
+type JsonValue = string | number | boolean | null | JsonObject | JsonArray;
+
+interface JsonObject {
+  [key: string]: JsonValue;
+}
+
+interface JsonArray extends Array<JsonValue> {}
+
+export async function install(): Promise<void> {
+  await Promise.all([
+    installDefault(),
+    installVariants(),
+  ]);
+}
+
+function _install(bowerJson?: JsonObject, directory?: string): Promise<void> {
+  return new Promise(async (resolve, reject) => {
+    const config = defaultBowerConfig({
+      save: false,
+      directory,
+    });
+    const cwd: string = config.cwd || process.cwd();
+    const renderer = new StandardRenderer('install', {
+      cwd,
+      color: true,
+    });
+    const bowerLogger = new BowerLogger();
+
+    const project = new BowerProject(config, bowerLogger);
+
+    bowerLogger.on('log', (log: bower.LogData) => {
+      renderer.log(log);
+    });
+    bowerLogger.on('end', (data: any) => {
+      renderer.end(data);
+      resolve();
+    });
+    bowerLogger.on('error', (err: Error) => {
+      renderer.error(err);
+      reject(err);
+    });
+
+    // This is the only way I could find to provide a JSON object to the
+    // Project. It's a hack, and might break in the future, but it works.
+    if (bowerJson) {
+      (project as any)._json = bowerJson;
+      (project as any)._jsonFile = path.join(cwd, 'bower.json');
+    }
+
+    project.install([], {save: false}, config);
+  });
+}
+
+async function installDefault(): Promise<void> {
+  logger.info(`Installing Bower components`);
+  await _install();
+  logger.info(`Finished installing Bower components`);
+}
+
+async function installVariants(): Promise<void> {
+  const json = await new Promise<any>((resolve, reject) => {
+    const config = defaultBowerConfig({
+      save: false,
+    });
+    const cwd = config.cwd || process.cwd();
+    readBowerJson(cwd, {}, (err: any, json: any) => {
+      err ? reject(err) : resolve(json);
+    });
+  });
+  const variants = json['variants'];
+  if (variants) {
+    await Promise.all(Array.from(Object.keys(variants)).map(async (variantName) => {
```

Comment:
Because JavaScript sucks and `Map.prototype.keys` returns an iterator, so I thought `Object.keys` would too, and iterators don't have `.map()`.